dashboard/app/helpers/users_helper.rb

```diff
@@ -1,49 +1,90 @@
 module UsersHelper
   include ApplicationHelper
 
-  # Summarize the current user's progress within a certain script.
+  # Summarize a user and his or progress progress within a certain script.
+  # Example return value:
+  # { "linesOfCode": 34, "linesOfCodeText": "Total lines of code: 34", "disableSocialShare": true,
+  #   "levels": {"135": {"status": "perfect", "result": 100}}}
   def summarize_user_progress(script, user = current_user, exclude_level_progress = false)
     user_data = {}
-    uls = {}
-    script_levels = script.script_levels
-    if user
-      lines = user.total_lines
+    merge_user_summary(user_data, user)
+    merge_script_progress(user_data, user, script)
+    user_data
+  end
 
-      uls = user.user_levels_by_level(script)
+  # Summarize a user and his or her progress across all scripts.
+  # Example return value:
+  # {
+  #     "lines": 34, "linesOfCodeText": "Total lines of code: 34", "disableSocialShare": true,
+  #     "scripts": {
+  #         "49": {
+  #             "name": "course2",
+  #             "levels": {"135": {"status": "perfect", "result": 100}}},
+  #         "46": {
+  #             "name": "artist",
+  #             "levels": {
+  #                "1138": {"status": "attempted", "result": 5},
+  #                "1147": {"status": "perfect", "result": 30}}}}}
+  def summarize_user_progress_for_all_scripts(user)
+    user_data = {}
+    merge_user_summary(user_data, user)
+    user_data[:scripts] = {}
+    merge_scripts_progress(user_data[:scripts], user)
+    user_data
+  end
+
+  # Merge the user summary into the specified result hash.
+  private def merge_user_summary(user_data, user)
+    if user
       user_data[:disableSocialShare] = true if user.under_13?
       user_data[:isTeacher] = true if user.teacher?
+      user_data[:linesOfCode] = user.total_lines
+    else
+      user_data[:linesOfCode] = client_state.lines
+    end
+    user_data[:linesOfCodeText] = I18n.t('nav.popup.lines', lines: user_data[:linesOfCode])
+    user_data
+  end
+
+  # Merge the progress for all scripts into the specified result hash.
+  private def merge_scripts_progress(user_data, user)
+    UserLevel.where(user_id: user.id).each do |ul|
+      script_id = ul.script_id
+      script = Script.get_from_cache(script_id)
+      user_data[script_id] ||= {name: script.name}
```

Comment:
I may be better to key this by `script.name` instead of `script_id`, unless the client needs to know the database IDs.